### PR TITLE
[DTensor] Fix silently skipping Partial related redistribution in backward

### LINKED ISSUE #167655: DebugMode reports incorrect redistribute placement in backwards
### 🐛 Describe the bug

Source: https://colab.research.google.com/drive/1zugXJm47oX4LOGfczk1H0xKs9d_Cn-sm#scrollTo=ahfnmPO3gtS3

Repro:

```
import torch
from torch import Tensor
from torch.distributed.tensor import (
    DTensor,
    DeviceMesh,
    distribute_tensor,
    init_device_mesh,
    Partial,
    Replicate,
    Shard,
)
from torch.distributed.tensor.placement_types import _StridedShard
from torch.distributed._local_tensor import (
    local_tensor_mode,
    LocalTensor,
    LocalTensorMode,
)
import traceback

S = Shard
R = Replicate()
_SS = _StridedShard

def product(it):
    x = 1
    for i in it:
        x *= i
    return x

def arange_nd(*sizes):
    if len(sizes) == 1 and isinstance(sizes[0], (list, tuple)):
        sizes = sizes[0]
    return torch.arange(product(sizes)).view(sizes)

def reconcile(l: Tensor):
    """Asserts that a LocalTensor is the same on all ranks, and returns the single Tensor."""
    if isinstance(l, LocalTensor):
        return l.reconcile()
    return l

def init_local_tensor_mode(world_size):
    from torch.distributed import _local_tensor
    if _local_tensor._LOCAL_TENSOR_MODE:
        for lm in list(reversed(_local_tensor._LOCAL_TENSOR_MODE)):
            lm.__exit__(None, None, None)
    try:
        torch.distributed.destroy_process_group()
    except AssertionError:
        pass
    torch.distributed.init_process_group(
        "fake",
        rank=0,
        world_size=world_size,
    )
    lm = LocalTensorMode(world_size)
    lm.__enter__()
    return world_size

# https://github.com/jax-ml/jax/pull/29381

world_size = init_local_tensor_mode(4)
mesh = init_device_mesh("cpu", (2, 2), mesh_dim_names=("x", "y"))

a = distribute_tensor(arange_nd(8, 4).float(), mesh, [S(0), S(1)]).requires_grad_()
b = distribute_tensor(arange_nd(4, 6).float(), mesh, [R, S(0)]).requires_grad_()
c = distribute_tensor(arange_nd(8, 4).float(), mesh, [S(0), S(1)]).requires_grad_()
d = distribute_tensor(arange_nd(4, 6).float(), mesh, [R, S(0)]).requires_grad_()
e = a @ b
f = c @ d
g = e + f
h = g.redistribute(placements=[S(0), R])  # comms!
i = h.sum()

e.retain_grad()
f.retain_grad()
g.retain_grad()
h.retain_grad()
i.retain_grad()

from torch.utils._debug_mode import DebugMode
with DebugMode(record_realtensor=False) as debug_mode:
    i.backward()
print(debug_mode.debug_string())

"""
print(a.grad.placements)
print(b.grad.placements)
print(c.grad.placements)
print(d.grad.placements)
print(e.grad.placements)
print(f.grad.placements)
print(g.grad.placements)
print(h.grad.placements)
print(i.grad.placements)
"""
```

This prints:

```
  aten::ones_like(dt: f32[]| PR, pin_memory=False, memory_format=torch.preserve_format)
  aten::clone(dt: f32[]| RR, memory_format=torch.contiguous_format)
  aten::expand(dt: f32[]| RR, [8, 6])
  aten::clone(dt: f32[8, 6]| RR, memory_format=torch.contiguous_format)
    redistribute_input(t: f32[8, 6], trace: RR->S(0)R->S(0)P)
  aten::clone(dt: f32[8, 6]| S(0)R, memory_format=torch.contiguous_format)
  aten::clone(dt: f32[8, 6]| S(0)R, memory_format=torch.contiguous_format)
  aten::t(dt: f32[8, 4]| S(0)S(1))
  aten::mm(dt: f32[4, 8]| S(1)S(0), dt: f32[8, 6]| S(0)R)
  aten::t(dt: f32[4, 6]| RS(0))
  aten::mm(dt: f32[8, 6]| S(0)R, dt: f32[6, 4]| RS(1))
  aten::detach(dt: f32[8, 4]| S(0)S(1))
  aten::detach(dt: f32[4, 6]| PS(0))
  aten::clone(dt: f32[8, 6]| S(0)R, memory_format=torch.contiguous_format)
  aten::t(dt: f32[8, 4]| S(0)S(1))
  aten::mm(dt: f32[4, 8]| S(1)S(0), dt: f32[8, 6]| S(0)R)
  aten::t(dt: f32[4, 6]| RS(0))
  aten::mm(dt: f32[8, 6]| S(0)R, dt: f32[6, 4]| RS(1))
  aten::detach(dt: f32[8, 4]| S(0)S(1))
  aten::detach(dt: f32[4, 6]| PS(0))
```

In particular, look at the redistribute:

```
    redistribute_input(t: f32[8, 6], trace: RR->S(0)R->S(0)P)
```

It claims that the tensor is redistributed to partial. But it is not, in DTensor, we have hard-coded that you should stay replicate when this happens.

```
        output = redistribute_local_tensor(
            local_tensor,
            current_spec,
            previous_spec,
            async_op=async_op,
            is_backward=True,
        )

        if output.dtype != ctx.original_dtype:
            output = output.to(ctx.original_dtype)

        # normalize the target placement to replicate if it is partial
        normalized_placements: list[Placement] = []
        for previous_placement in previous_spec.placements:
            if previous_placement.is_partial():
                # keep target placement to replicate instead of partial in this case
                normalized_placements.append(Replicate())
            else:
                normalized_placements.append(previous_placement)

```

But because the redistribute hook is in redistribute_local_tensor it reports the P placement (prior to it getting normalized).

The "right" fix is to stop hacking this up with the normalization after the fact, and ask for the redistribute we actually want.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @zpcore 

### Versions

main

### PR DESCRIPTION
**Pull request by zpcore**
### Summary
This PR removes redistribution "hacks" in `Redistribute.backward` that silently skips Partial related operation. Instead, we let `redistribute_local_tensor` control whether to do the hack / optimization or not. If we do the optimization, `redistribute_local_tensor` will 
- return the final updated placements (where Partial will mostly be getting rid of during backward), and
- update `torch.utils._debug_mode` with the real _TransformInfo executed.
- even we set `is_backward=False` for backward path, the result should also be correct. We just skip the optimization in this case.

The idea is simple, we update `redistribute_local_tensor` to return both the redistributed tensor and the corresponding placement of the returned tensor:
```
def redistribute_local_tensor(..., is_backward: bool = False, ,...) -> tuple[torch.Tensor, list[Placement]]:
```

### Background
If we go through this part of code https://github.com/pytorch/pytorch/blob/a344069f2aba6a87c0ab3fab488e83b80691457f/torch/distributed/tensor/_redistribute.py#L921-L939
- redistribute_local_tensor function silently skips the redistribution when we redistribute R->P in the backward.
- But also in the backward, we update placement hint of Partial into Replicate.  

Those two "hacks" make the code hard to maintain. Also it confuses users that the real redistribution doesn't follow the computed redistribution path.

Fixes #167655

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk

### CODE REVIEW

File: torch/distributed/tensor/_redistribute.py [hunk 1]
```diff
@@ -918,29 +999,19 @@ def backward(ctx, grad_output: "dtensor.DTensor"):  # type: ignore[override]
             local_tensor = grad_output._local_tensor
             current_spec = grad_output._spec
 
-        output = redistribute_local_tensor(
+        output, end_placements = redistribute_local_tensor(
```
Comment (zpcore) [inline torch/distributed/tensor/_redistribute.py:1032]: Here the optimization is all managed inside redistribute_local_tensor function.

Comment (ezyang) [reply torch/distributed/tensor/_redistribute.py:1032]: It is possible this strategy is the optimal one, but I need to ask you about an alternative. redistribute_local_tensor takes in a placement which you want to redistribute to. Why do we not just have the code HERE compute the /actual/ new placement we want to go to (and not the "incorrect" one), i.e., why is it necessary for redistribute_local_tensor to tell us what the actual end placements is.

I would be more sympathetic to this PR if the code change was short, but there's also tons of fuzz in lots of other places that are very suspicious!!

**Review by ezyang — CHANGES_REQUESTED** [review]
blocking prior to resolution of comment

File: torch/distributed/tensor/_redistribute.py (hunk 1 above)
Comment (zpcore) [reply torch/distributed/tensor/_redistribute.py:1032]: Let me try to convince you...

---
This is not the only place that need to handle the `is_backward=True`. If we want to relocate the optimization logic from `redistribute_local_tensor` here into the backward(), we also need to update other backward places that sets `is_backward=True`.  Of course we can argue that `_maybe_adjust_transform_info_inplace` can be called directly from those backward(), but the function input depends on the output from `_gen_transform_infos`.

---
The change is not that fuzz. The `redistribute_local_tensor` logic is the same as before. I basically split the `redistribute_local_tensor` out into three parts: 
1) `_gen_transform_infos`: generates transform_infos.
2) `_maybe_adjust_transform_info_inplace`: checks if we want to update the target placements with some optimizations (like special handling P->R in the backward) based on transform_infos from 1).
3) `_execute_transform`:  Calls collective ops based on generated redistribution steps from 2) to update tensor data.  

Original code mixes 2) and 3) together.

File: torch/distributed/tensor/_redistribute.py [hunk 2]
```diff
@@ -723,12 +908,19 @@ def redistribute_local_tensor(
             current_spec, target_spec, use_graph_based_transform
         )
 
+    end_placements = _maybe_adjust_transform_info_inplace(
```
Comment (vishal9-team) [inline torch/distributed/tensor/_redistribute.py:911]: nit:can this be of the form, 

`adjusted_transform_infos, final_placements = _maybe_adjust_transform_info(...)`

File: torch/distributed/tensor/_redistribute.py (hunk 1 above)
Comment (ezyang) [reply torch/distributed/tensor/_redistribute.py:1032]: There's only one substantive other call of `is_backward` right? In fact, that other call does NOT renormalize the placements when it's done. Why? This seems suspicious!
